File: Kth Smallest Element in a BST - Leetcode 230.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def kthSmallest(self, root: Optional[TreeNode], k: int) -> int:
        k_ans = [k]
        ans = [0]

        def dfs(node):
            if not node:
                return
            
            dfs(node.left)
            k_ans[0] = k_ans[0] - 1

            if k_ans[0] == 0:
                ans[0] = node.val

            if k > 0:
                dfs(node.right)
            
        dfs(root)
        return ans[0]
        # Time: O(n)
        # Space: O(n)

File: Kth Smallest Element in a BST - Leetcode 230/Kth Smallest Element in a BST - Leetcode 230.java
import java.util.*;

public class Solution {
    private int count;
    private int ans;

    public int kthSmallest(TreeNode root, int k) {
        count = k;
        ans = 0;
        dfs(root);
        return ans;
    }

    private void dfs(TreeNode node) {
        if (node == null) {
            return;
        }

        dfs(node.left);

        if (count == 1) {
            ans = node.val;
        }

        count--;
        if (count > 0) {
            dfs(node.right);
        }
    }
}
